File: python/boj16234.py
import sys, copy
input = lambda: sys.stdin.readline().rstrip()
        
def bfs():
    q1 = len(q[-1])
    q2 = q1
    while q2:
        q1 = len(q[-1])
        q2 = q1
        for i in range(q1):
            x, y, _ = q[-1][i]
            for i in range(4):
                temp_x = x+dx[i]
                temp_y = y+dy[i]
                if 0 <= temp_x < N and 0 <= temp_y < N:
                    if L <= abs(A[x][y] - A[temp_x][temp_y]) <= R and not vA[temp_x][temp_y]:
                        q[-1].append((temp_x, temp_y, A[temp_x][temp_y]))
                        vA[temp_x][temp_y] = True
                        q2 += 1
            q2 -= 1
        


N, L, R = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(N)]
dx = [1,0,-1,0]
dy = [0,1,0,-1]

for m in range(2000):
    q = []
    vA = [[False for _ in range(N)] for _ in range(N)]

    for i in range(N):
        for j in range(N):
            if not vA[i][j]:
                vA[i][j] = True
                q.append([(i,j,A[i][j])])
                bfs()
            if len(q[-1]) == 1:
                q.pop(-1)

    if len(q) == 0:
        break

    for unite in q:
        _sum = 0
        for i in range(len(unite)):
            _sum += unite[i][2]
        avg = _sum//len(unite)
        for x, y, _ in unite:
            A[x][y] = avg

print(m)

File: cpp/boj16234.cpp
#include <iostream>
using namespace std;

int N,L,R, A[51][51];
int dx[]={1,0,-1,0}, dy[]={0,1,0,-1};
int num_mem, _sum;

void findassociation(int a[][51], int x, int y, int num, bool *is){
    int tmpx, tmpy;
    for(int i(0); i< 4; i++){
        tmpx = x+dx[i]; tmpy=y+dy[i];
        if(0<=tmpx && tmpx < N && 0<=tmpy && tmpy < N){
            if(L <= abs(A[x][y]-A[tmpx][tmpy])&&abs(A[x][y]-A[tmpx][tmpy]) <=R){
                if(!a[tmpx][tmpy]){
                    a[tmpx][tmpy] = num;
                    _sum += A[tmpx][tmpy]; num_mem++;
                    *is = true;
                    findassociation(a, tmpx, tmpy, num, is);
                }
            }
        }
    }
}

bool populationchange(){
    int asso[51][51]={{0,}}, num_asso=1, _avg[50*50];
    bool is_change=false;
    for(int n(0); n < N; n++){
        for(int m(0); m < N; m++){
            if(!asso[n][m]){
                bool is_asso=false;
                num_mem=1; _sum=A[n][m];
                asso[n][m] = num_asso;
                findassociation(asso,n,m,num_asso, &is_asso);
                if(is_asso) {
                    _avg[num_asso++] = _sum/num_mem;
                    is_change=true;
                }
                else asso[n][m] = 0;
            }
        }
    }
    for(int n(0); n < N; n++){
        for(int m(0); m < N; m++){
            if(asso[n][m]){
                A[n][m] = _avg[asso[n][m]];
            }
        }
    }

    return is_change;
}

int main(){
    freopen("input.txt", "r", stdin);
    cin >> N >> L >> R;
    for(int n(0); n < N; n++){
        for(int m(0); m < N; m++){
            cin >> A[n][m];
        }
    }
    
    int day(0); bool change=true;
    while(change){
        change = populationchange();
        if(change) day++;
    }

    cout << day;
    fclose(stdin);
    return 0;
}